lib/relooper/src/graph/cfg/mod.rs

```diff
@@ -0,0 +1,175 @@
+        self.map_label(|l| l)
+    }
+
+    pub fn nodes(&self) -> HashSet<&T> {
+        self.out_edges.keys().collect()
+    }
+
+    pub fn children(&self, label: &T) -> HashSet<&T> {
+        self.out_edges
+            .get(label)
+            .into_iter()
+            .flat_map(|edge| edge.to_vec())
+            .collect()
+    }
+
+    pub fn add_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let nodes = self.nodes();
+        // todo unncecessary collect
+        for n in edge
+            .to_vec()
+            .into_iter()
+            .filter(|n| !nodes.contains(n))
+            .collect::<Vec<_>>()
+        {
+            // todo clone
+            self.add_node(n.clone());
+        }
+
+        let prev_edge = self.out_edges.insert(from, edge);
+        match prev_edge {
+            None | Some(Terminal) => {}
+            _ => panic!("adding edge over already present one"),
+        }
+    }
+
+    pub fn add_node(&mut self, n: T) {
+        self.out_edges.insert(n, Terminal);
+    }
+
+    pub fn remove_edge(&mut self, from: T, edge: CfgEdge<T>) {
+        let removed_edge = self.out_edges.remove(&from);
+        assert!(removed_edge == Some(edge));
+    }
+
+    pub fn add_edge_or_promote(&mut self, from: T, to: T) {
+        match self.out_edges.remove(&from) {
+            None | Some(Terminal) => self.out_edges.insert(from, Uncond(to)),
+            Some(Uncond(uncond)) => self.out_edges.insert(from, Cond(to, uncond)),
+            _ => panic!("edge (should be absent) or (shouldn't be `Cond`)"),
+        };
+    }
+
+    pub fn edge(&self, label: &T) -> &CfgEdge<T> {
+        self.out_edges
+            .get(label)
+            .expect("any node should have outgoing edges")
+    }
+}
+
+impl<TLabel: Eq + Hash + Copy> Cfg<TLabel> {
+    pub fn from_edges(entry: TLabel, edges: &HashMap<TLabel, CfgEdge<TLabel>>) -> Self {
+        let mut cfg = Cfg::new(entry);
+        for (&from, &edge) in edges.iter() {
+            cfg.add_edge(from, edge);
+        }
+
+        cfg
+    }
+
+    #[deprecated]
+    pub fn from_vec(entry: TLabel, edges: &[(TLabel, CfgEdge<TLabel>)]) -> Self {
+        let edges_map: HashMap<TLabel, CfgEdge<TLabel>> = edges.iter().copied().collect();
+        Self::from_edges(entry, &edges_map)
+    }
+}
+
+impl<TLabel: CfgLabel> Cfg<TLabel> {
+    pub fn in_edges(&self) -> HashMap<TLabel, HashSet<TLabel>> {
+        let mut in_edges: HashMap<TLabel, HashSet<TLabel>> = HashMap::default();
+
+        for (&from, to_edge) in &self.out_edges {
+            for &to in to_edge.to_vec() {
+                in_edges.entry(to).or_default().insert(from);
+            }
+        }
+
+        in_edges
+    }
+
+    fn reachable_nodes(&self) -> HashSet<&TLabel> {
+        Bfs::start_from(&self.entry, |label| self.children(label)).collect()
+    }
+
+    pub fn strip_unreachable(&mut self) {
+        let nodes: HashSet<TLabel> = self.nodes().into_iter().copied().collect(); // TODO get rid of copies
```

Comment:
Remaining todo. This one should also be easy to fix. Remove the call to `copied` and change the type to `HashSet<&TLabel>` should work.

Reply:
Unfortunately it's is not trivial problem with borrow checker, I think now is better just left it as is

Reply:
I suggest rewriting the function as follows:

```rust
    pub fn strip_unreachable(&mut self) {
        let unreachable_nodes: HashSet<TLabel> = self
            .nodes()
            .difference(&self.reachable_nodes())
            .into_iter()
            .map(|n| **n)
            .collect();
        for unreachable in unreachable_nodes {
            self.out_edges.remove(&unreachable);
        }
    }
```

The reason we cannot avoid doing at least one copy is because we need to use the same `out_edges` collection for all the operations here. If we do not copy the `TLabel` data out of the collection then we cannot remove it (removing from the collection would invalidate the pointers we created while traversing over `out_edges` in the first place). But in the way I have refactored it we minimize how many nodes need to be copy. We only copy after taking the difference instead of copying all the nodes twice.

Reply:
Thanks! I rewrite it [here](https://github.com/aurora-is-near/evm2near/pull/9/commits/c8c5aa752c29308f4ccd130834ebb8b88f979377)

Reply:
Thanks @MCJOHN974 . The change was accidentally reverted in [this commit](https://github.com/aurora-is-near/evm2near/pull/9/commits/0c07fd79731e7e3b7bfd86bfaff07fc2c64a8200). Please push the rewrite again.